Added device plugin e2e kubelet failure test

RenaudWasTaken opened the pull request:
Signed-off-by: Renaud Gaubert <renaud.gaubert@gmail.com>

**What this PR does / why we need it**:
This is part of issue #52859 (fixes #52859)

This PR adds a e2e_node test for the device plugin.
Specifically it implements testing of failure handling by the device plugin components in case Kubelet restart / crashes.

I might try to refactor the GPU tests in a later PR.

**Special notes for your reviewer**:
@jiayingz @vishh 

**Release note**:
```release-note
NONE
```

RenaudWasTaken committed:
Added device plugin e2e kubelet failure test

Signed-off-by: Renaud Gaubert <renaud.gaubert@gmail.com>

dims commented:
/ok-to-test

RenaudWasTaken commented:
/retest

RenaudWasTaken commented:
/test pull-kubernetes-e2e-gce-gpu
--

RenaudWasTaken commented:
/test pull-kubernetes-e2e-gce-gpu

RenaudWasTaken commented:
/test pull-kubernetes-federation-e2e-gce

RenaudWasTaken commented:
/retest

RenaudWasTaken commented:
/retest

cmluciano commented on test/e2e/scheduling/device-plugin.go:
```diff
@@ -0,0 +1,110 @@
+
+	BeforeEach(func() {
+		if !isClusterRunningCOS(f) {
+			Skip("Nvidia GPU tests are supproted only on Container Optimized OS image currently")
```
s/supproted/supported

Nvidia GPU tests are currently only supported on the COS image

cmluciano commented on test/e2e/scheduling/device-plugin.go:
```diff
@@ -0,0 +1,110 @@
+		}
+	})
+
+	It("checks that when Kubelet restarts exclusive GPU assignation to pods is kept. Runs on Container Optimized OS only", func() {
```
Don't think we need the COS note here too

cmluciano commented on test/e2e/scheduling/device-plugin.go:
```diff
@@ -0,0 +1,110 @@
+
+		By("Checking that pods got a different GPU")
+		Expect(uuid1).To(Not(Equal(uuid2)))
+	})
```
Are the containers cleaned up after this test?

RenaudWasTaken replied:
Good catch. That would explain why the tests are failing.
I think the other device plugin test conflicts with this one

RenaudWasTaken commented:
@jiayingz what do you think of this test ?
Also I've noticed that GPU tests keep failing on driver installation should I increase the timeout ?

cmluciano reviewed (commented):
Few nits and note on resource cleanup

vishh commented:
This test can be part of node e2e since it's only testing a single node. Running with 6 GPUs in cluster mode (3 nodes per cluster by default) instead of running with 2 GPUs for every test run is preferred.

RenaudWasTaken commented:
> This test can be part of node e2e since it's only testing a single node. Running with 6 GPUs in cluster mode (3 nodes per cluster by default) instead of running with 2 GPUs for every test run is preferred.

Agreed this was actually a question I had. Should I extend this test to all nodes or just move it to node e2e.
Moving to node e2e it is then!

timothysc commented:
/cc @derekwaynecarr may have more opinions here.

timothysc commented on test/e2e/scheduling/device-plugin.go:
```diff
@@ -0,0 +1,110 @@
+	. "github.com/onsi/gomega"
+)
+
+var _ = SIGDescribe("[Feature:GPU]", func() {
```
I kind of against this being a e2e unless it can be more generally applied across (N) environments.

RenaudWasTaken replied:
This is the current behavior (and has been for a while now) of the e2e nvidia gpus test.
And it will probably not change in the foreseeable future either.

Agreed that it would be better if it could be applied across multiple environment :)

RenaudWasTaken replied:
Yep I noticed that an hour ago when moving this to node_e2e.

> That may require us to have a fake device plugin implementation just for testing purpose. I think that will probably fall into 1.9 release timeline

I think so too

> For now, maybe we can add this test into the existing test/e2e/scheduling/nvidia-gpus.go given that it is nvidia gpu specific?

We're moving this to node_e2e as per @vishh's suggestion

RenaudWasTaken commented on test/e2e/scheduling/device-plugin.go:
```diff
@@ -0,0 +1,110 @@
+
+	BeforeEach(func() {
+		if !isClusterRunningCOS(f) {
+			Skip("Nvidia GPU tests are supproted only on Container Optimized OS image currently")
```
Can you develop what you mean here?

jiayingz reviewed (commented):
Thanks a lot for taking this effort, Renaud! Please see my inline comments.

RenaudWasTaken commented on test/e2e/scheduling/device-plugin.go:
```diff
@@ -0,0 +1,110 @@
+		spec = makeCudaAdditionDevicePluginTestPod()
+		spec.Spec.NodeName = n.Name
+
+		p2 := f.PodClient().Create(spec)
```
I don't know why I thought `f.PodClient().WaitForSuccess(p2.Name, 5*time.Minute)` meant that this waited for the pod to be running...

Thanks for noticing this

RenaudWasTaken commented on test/e2e/scheduling/device-plugin.go:
```diff
@@ -0,0 +1,110 @@
+	})
+})
+
+func installDriver(f *framework.Framework) {
```
Copied the wrong link here...
Thanks for noticing this

RenaudWasTaken commented:
Thanks for the comments. Will publish a new version by tomorrow :)

RenaudWasTaken commented:
@vishh @jiayingz updated the following:
- Moved to e2e_node (updated the calls accordingly)
- Cleans up pods after each tests
- Addressed nits
- Moved some functions shared to framework/utils

RenaudWasTaken commented:
/test pull-kubernetes-e2e-gce-bazel

mindprince commented on test/e2e/framework/util.go:
```diff
@@ -187,6 +188,13 @@ const (
+
+	// GPUResourceName is the extended name of the GPU resource since v1.8
+	// this uses the device plugin mechanism
+	NVIDIAGPUResourceName = "nvidia.com/gpu"
```
Let's not add this here. We don't want to dictate what nvidia chooses as its resource name for device plugins.

RenaudWasTaken replied:
I agree  that we don't want to dictate what nvidia chooses as its resource name. 
However, right now this the resource name advertised by the device plugin @jiayingz implemented.

Do you have another suggestion :) ?

mindprince replied:
I would move these methods out from framework/util.go and put them back where they are getting used. As far as I understand testing based on a vendor specific device plugin is temporary. Eventually we will use a dummy device plugin to test that the device plugin functionality in kubelet works as expected. And each vendor would have their own tests for device plugins.

RenaudWasTaken replied:
> As far as I understand testing based on a vendor specific device plugin is temporary

Right. However I don't want to duplicate these methods as this is now part of e2e_node and they were used in e2e/scheduling.

> Eventually we will use a dummy device plugin to test that the device plugin functionality in kubelet works as expected

As mentioned by @jiayingz I think we still want these to be used in a real scenario such as NVIDIA GPUs.

RenaudWasTaken replied:
Just to add on the duplication and usage part. I think it makes more sense to centralize this under framework/utils.go as this means that if NVIDIA wants to change this name or google the COS name advertised then it won't have to go change that occurrence everywhere

jiayingz replied:
Can we just add DsFromManifest() in this file and keep other definitions local in test/e2e_node/device-plugin.go? It is ok to have some dup code for now. We can add a common pkg later if we see a lot duplicate functionalities.

RenaudWasTaken replied:
I'm sorry but I seem to fail seeing the benefits of having this code be duplicated in two separate files vs keeping it centralized in a place that seem easier to change.
Is there anything I am missing ?

jiayingz replied:
If you feel strong about avoiding duplicate code, can we put them into gpu_util.go? I think our main concern is these definitions are gpu specific and irrelevant to most people using e2e framework.

RenaudWasTaken replied:
I can put them in gpu_util.go I'll also add some comments :)
I think that should solve everyone's concerns

mindprince commented on test/e2e/framework/util.go:
```diff
@@ -5015,3 +5023,50 @@ func DumpDebugInfo(c clientset.Interface, ns string) {
+
+// NumberOfGPUs returs the number of GPUs advertised by a node
+// This is based on the Device Plugin system
+func NumberOfGPUs(node *v1.Node) int64 {
```
This also shouldn't be added here. GPUs != capacity[NVIDIAGPUResourceName] when there are other GPU devices etc.

RenaudWasTaken replied:
We can call it NumberOfNVIDIAGPUs

rohitagarwal003 reviewed (changes requested):
Base this test on `test/e2e_node/gpus.go` instead of `test/e2e/scheduling/nvidia-gpus.go`.

RenaudWasTaken commented:
/retest

RenaudWasTaken commented:
@mindprice what do you mean when you say:
> Base this test on test/e2e_node/gpus.go instead of test/e2e/scheduling/nvidia-gpus.go.

Do you have specific examples ?

RenaudWasTaken commented:
/test pull-kubernetes-e2e-gce-bazel

RenaudWasTaken commented:
/retest

jiayingz commented on test/e2e_node/device-plugin.go:
```diff
@@ -0,0 +1,152 @@
+
+		By("Restarting Kubelet and creating another GPU pod")
+		restartKubelet(f)
+
```
Here can we also add a case to restart p1 and verifies that it gets the same device after restarts?

jiayingz commented on test/e2e_node/device-plugin.go:
```diff
@@ -0,0 +1,152 @@
+}
+
+// TODO: Find a uniform way to deal with systemctl/initctl/service operations. #34494
+func restartKubelet(f *framework.Framework) {
```
Given that we only expect the test to run on cos currently, do we still need to support these operations?

RenaudWasTaken replied:
Don't have any opinion on this. If I had to choose it doesn't seem like a bad solution to support all three init systems and this function can be reused in other tests (and the function is already written).

What init system does COS use ?

jeremyeder replied:
@jiayingz "only expect the test to run on cos currently" -- why?

jiayingz replied:
Currently we only have a single device plugin implementation, i.e., cos based nvidia gpu device plugin. For 1.8 time frame, it seems the easiest way to test the end-to-end device plugin behavior with this reference implementation. In 1.9, we do hope to make the device plugin e2e test generic by using a stub device plugin implementation. I would also hope we may generalize some of the functions used here into a pkg so that different specific device plugin implementations can leverage it in their own e2e testing in separate repositories.

timothysc commented on test/e2e_node/device-plugin.go:
```diff
@@ -0,0 +1,152 @@
+	// Install NVIDIA driver if cluster is running COS
+	BeforeEach(func() {
+		if !framework.IsNodeRunningCOS(getLocalNode(f)) {
+			Skip("NVIDIA GPU tests are supported only on Container Optimized OS image currently")
```
What I was saying before is I *do not* think it is wise to add this gate to an e2e test.  Folks will want to enable this feature outside of GCE/GKE and we can warn, but we should not gate on it. 

/cc @ConnorDoyle @derekwaynecarr @davidopp @jeremyeder

RenaudWasTaken replied:
In that case the problem is not related to this PR only but all the GPU related tests (which are already merged in master) and I suggest we open an issue to discuss this as the scope of this discussion (which seem to be installing NVIDIA drivers from a container) is extremely vast.

That would allow this PR to continue being merged as it is doing the same kind of operations that are already in the code base (unless you consider this a blocker) :)

Does this seem like a reasonable course of action @timothysc ?

jeremyeder replied:
Hmm, test dependence on a single distro seems like a blocker to me.

RenaudWasTaken replied:
Currently NVIDIA does not officially support installing the NVIDIA drivers through containers (COS is a work in progress).

For 1.8 we are trying to e2e test the device plugin on one of it's target hardware (GPUs)  As mentioned above we don't have a unified or supported solution for installing the NVIDIA driver.
So right now we are limiting ourselves to GKE using Google's GPU device plugin + driver installer.

Note that tests running specifically on COS have already been in the code for 5 months see:
https://github.com/kubernetes/kubernetes/pull/45136

jiayingz replied:
Instead of calling this file device-plugin.go, can we just call it gpu-device-plugin.go to indicate its current scope? After 1.8, we may want to replace these specific functions with more generic interface functions, e.g., instead of IsNodeRunningCOS, we can just use device-plugin-framework.NodeOSChecking().

mindprince commented:
> what do you mean when you say:

> > Base this test on test/e2e_node/gpus.go instead of test/e2e/scheduling/nvidia-gpus.go.

> Do you have specific examples ?

This test is disruptive (restarts the kubelet etc.), so we don't want this test running in parallel with the other tests. The test in `test/e2e_node/gpus.go` is also disruptive, that's why we added `[Serial]` to its description, so that it runs in the suite which runs tests serially.

That file also doesn't check whether it's running on a particular distro: all it checks is whether the gpus are available on the node or not.
```
By("ensuring that Nvidia GPUs exist on the node")
if !checkIfNvidiaGPUsExistOnNode() {
        Skip("Nvidia GPUs do not exist on the node. Skipping test.") 
}
```
Which image type has these gpus are controlled from: `test/e2e_node/jenkins/image-config-serial.yaml`.
And the installation of the driver is in: `test/e2e_node/jenkins/gci-init-gpu.yaml`.

RenaudWasTaken commented:
Hi @mindprice!

> This test is disruptive (restarts the kubelet etc.), so we don't want this test running in parallel with the other tests. The test in test/e2e_node/gpus.go is also disruptive, that's why we added [Serial] to its description, so that it runs in the suite which runs tests serially.

Thanks for noticing this! It's a really good catch :)

> That file also doesn't check whether it's running on a particular distro: all it checks is whether the gpus are available on the node or not.

> Which image type has these gpus are controlled from: test/e2e_node/jenkins/image-config-serial.yaml.
And the installation of the driver is in: test/e2e_node/jenkins/gci-init-gpu.yaml.

Just talked with @jiayingz and I agree. This means that testing driver installation is basically useless in e2e_node. Will remove this.

However we do want to keep the number of gpu check so that we don't break on setups without GPUs :)

Thanks for the help with this!

RenaudWasTaken commented:
/test pull-kubernetes-e2e-gce-bazel

jiayingz commented on test/e2e_node/gpu-device-plugin.go:
```diff
@@ -0,0 +1,130 @@
+)
+
+// Serial because the test restarts Kubelet
+var _ = framework.KubeDescribe("[Feature:GPUDevicePlugin][Serial]", func() {
```
Do we need to have a separate yaml file like test/e2e_node/jenkins/gci-init-gpu.yaml?

jiayingz replied:
Never mind. I realized the test/e2e_node/jenkins/gci-init-gpu.yaml config is used by all tests running in serial.

RenaudWasTaken commented:
/retest

jiayingz commented on test/e2e_node/gpu-device-plugin.go:
```diff
@@ -0,0 +1,130 @@
+	})
+
+	It("checks that when Kubelet restarts exclusive GPU assignation to pods is kept.", func() {
+		n := getLocalNode(f)
```
Do we also need to enable device plugin feature gate before running the test?

RenaudWasTaken commented:
/retest

jiayingz commented on test/e2e_node/gpu-device-plugin.go:
```diff
@@ -0,0 +1,130 @@
+	BeforeEach(func() {
+		if framework.NumberOfNVIDIAGPUs(getLocalNode(f)) < 2 {
+			Skip("Not enough GPUs to execute this test (at least two needed)")
+		}
```
Without device plugin feature gate enabled and the daemonset pod running, this test will fail.

jiayingz commented on test/e2e_node/gpu-device-plugin.go:
```diff
@@ -0,0 +1,130 @@
+
+	It("checks that when Kubelet restarts exclusive GPU assignation to pods is kept.", func() {
+		n := getLocalNode(f)
+
```
I think we still need to run the device plugin daemonset pod before running gpu pod. The container we run in gci-init-gpu.yaml just installs the driver.

RenaudWasTaken commented:
/retest

mindprince commented on test/e2e/framework/gpu_util.go:
```diff
@@ -0,0 +1,63 @@
+}
+
+// NVIDIADevicePlugin returns the official Google Device Plugin pod for NVIDIA GPU in GKE
+func NVIDIADevicePlugin(ns string) *v1.Pod {
```
I would insist that we don't put these definitions in the framework package.

Both the methods: `NumberOfNVIDIAGPUs()` and `NVIDIADevicePlugin()` are used only in `e2e_node/gpu-device-plugin.go` and can live there.

RenaudWasTaken replied:
`NumberOfNVIDIAGPUs()` is only used in `e2e_node/gpu-device-plugin.go` mostly because I don't want to make use of it in e2e/scheduling/nvidia-gpu` in this PR but in a later PR.

To answer your other concern, I do believe that we need a utils file that groups the utils functions used in the NVIDIA tests as they won't be used in one single package across the e2e code base.

The package `framework` seems to be doing exactly that (regroup all the `utils` function) which is why it seems to me that it's the right place.

@jiayingz mentioned that:
> Our main concern is these definitions are gpu specific and irrelevant to most people using e2e framework.

But that argument can be applied to a lot of utils functions the framework package is exposing (see pv_utils, rc_utils, nodes_utils, perf_utils, utils, exec_utils, statefulset_utils, firewall_utils).
Granted that some make more sense than others (API utils vs platform specific utils) but gpu_utils doesn't feel completely out of place when looking at a lot of these utils functions.

RenaudWasTaken commented:
/retest

RenaudWasTaken commented:
/retest

jiayingz committed:
Modified test/e2e_node/gpu-device-plugin.go to make sure it passes.

jiayingz commented:
@RenaudWasTaken Could you take a look at https://github.com/jiayingz/kubernetes/commit/ff423d0a2b8c6c9075cfba56f0c09d33d0195fd6. It is to be applied on top of your current PR. I tested it with 'make test-e2e-node REMOTE=true FOCUS="GPUDevicePlugin" SKIP="" IMAGE_CONFIG_FILE=test/e2e_node/jenkins/image-config-serial.yaml TEST_ARGS=--feature-gates=DynamicKubeletConfig=true PARALLELISM=1'. By default, make test-e2e-node doesn't run Serial tests, so you need to use this special command.

jiayingz commented on test/e2e_node/gpu-device-plugin.go:
```diff
@@ -0,0 +1,173 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Maybe rename this file as gpu_device_plugin.go to be consistent with the other files in this directory?

jiayingz reviewed (commented):
Just a nit. Otherwise lgtm.

jiayingz commented:
/assign @vishh

RenaudWasTaken commented:
/test pull-kubernetes-unit

vishh commented on test/e2e/framework/gpu_util.go:
```diff
@@ -0,0 +1,73 @@
+	NVIDIAGPUResourceName = "nvidia.com/gpu"
+
+	// GPUDevicePluginDSYAML is the official Google Device Plugin Daemonset NVIDIA GPU manifest for GKE
+	GPUDevicePluginDSYAML = "https://raw.githubusercontent.com/GoogleCloudPlatform/container-engine-accelerators/master/device-plugin-daemonset.yaml"
```
TODO for future: Parametrize it by making it a feature in TestFramework. That way we can override the daemonset in other setups (non COS).

vishh replied:
Bonus points for opening an issue about this.

vishh commented on test/e2e/framework/gpu_util.go:
```diff
@@ -0,0 +1,73 @@
+)
+
+// NumberOfGPUs returs the number of GPUs advertised by a node
+// This is based on the Device Plugin system and expected to run on a COS based node
```
COS is a requirement only because of the device plugin we use. Add a TODO to make it generic with a link to an issue ideally.

vishh commented on test/e2e_node/gpu-device-plugin.go:
```diff
@@ -0,0 +1,173 @@
+
+// TODO: Find a uniform way to deal with systemctl/initctl/service operations. #34494
+func restartKubelet(f *framework.Framework) {
+	stdout, err := exec.Command("sudo", "systemctl", "list-units", "kubelet*", "--state=running").CombinedOutput()
```
Why not do a `pkill kubelet`?

vishh reviewed (approved):
/lgtm
/approve

RenaudWasTaken commented:
/test pull-kubernetes-node-e2e

RenaudWasTaken commented:
/test pull-kubernetes-e2e-gce-bazel

jiayingz commented:
/lgtm

jiayingz commented:
/retest

mindprince commented:
@RenaudWasTaken This test is failing: https://k8s-testgrid.appspot.com/wg-resource-management#kubelet-serial-gce-e2e

```
/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e_node/gpu_device_plugin.go:107
Test Panicked
/usr/local/go/src/runtime/panic.go:489
```

RenaudWasTaken commented:
@mindprince current best guess with @jiayingz is that [this line](https://github.com/kubernetes/kubernetes/blob/cd37a28fa5165b46350df9e8dcd79539e610e48b/test/e2e_node/gpu_device_plugin.go#L160) is failing.

I'm thinking that after a kubelet reboot it's possible that `ContainerStatuses` might be empty.
